web/src/components/vis/VolcanoPlot.vue

```diff
@@ -23,106 +46,160 @@ export default {
     },
   },
 
-  data() {
-    return {
-      margin: {
-        top: 20,
-        right: 20,
-        bottom: 50,
-        left: 50,
-      },
-      radius: 3,
-      duration: 200,
-      xlabel: 'log2(Fold Change)',
-      ylabel: '-log10(p-value)',
-    };
-  },
+  setup(props) {
+    const mainRef = ref(document.createElement('div'));
 
-  computed: {
-    transformedRows() {
-      return this.rows.map((row) => ({
-        ...row,
-        x: row.log2FoldChange,
-        y: -Math.log10(row.pValue),
-      }));
-    },
-    xrange() {
-      const max = this.transformedRows.reduce((acc, d) => Math.max(acc, Math.abs(d.x)), 0);
+    const width = ref(800);
+    const height = computed(() => (
+      (props.rows.length * 10)
+      + (margin.top + margin.bottom)));
+
+    const {
+      dwidth,
+      dheight,
+      axisPlot,
+    } = useAxisPlot({
+      margin,
+      width,
+      height,
+      topAxis: false,
+    });
+
+    const transformedRows = computed(() => props.rows.map((row) => ({
+      ...row,
+      x: row.log2FoldChange,
+      y: -Math.log10(row.pValue),
+    }))) as Ref<TransformedRow[]>;
+
+    const xrange = computed(() => {
+      const max = transformedRows.value
+        .reduce((acc: number, d: TransformedRow) => Math.max(acc, Math.abs(d.x)), 0);
       return [-max, max];
-    },
-    yrange() {
-      const max = this.transformedRows.reduce((acc, d) => Math.max(acc, d.y), 0);
+    }) as Ref<[number, number]>;
+
+    const yrange = computed(() => {
+      const max = transformedRows.value
+        .reduce((acc: number, d: TransformedRow) => Math.max(acc, d.y), 0);
       return [0, max];
-    },
-  },
-  watch: {
-    rows() {
-      this.update();
-    },
-    minFoldChange() {
-      this.update();
-    },
-    minLogP() {
-      this.update();
-    },
-  },
-  mounted() {
-    this.update();
-  },
-  methods: {
-    update() {
-      // Compute the total variance in all the PCs.
-      const svg = select(this.$refs.svg);
-      this.axisPlot(svg);
+    }) as Ref<[number, number]>;
+
+    const scaleX = computed(() => scaleLinear()
+      .domain(xrange.value)
+      .range([0, dwidth.value]));
+
+    const scaleY = computed(() => scaleLinear()
+      // @ts-ignore d3 typings are bad
+      .domain(yrange.value)
+      // @ts-ignore d3 typings are bad
+      .range([dheight.value, 0]));
+
+    const axisX = computed(() => axisBottom(scaleX.value));
+    const axisY = computed(() => axisLeft(scaleY.value));
+
+    function onResize() {
+      const bb = mainRef.value.getBoundingClientRect();
+      width.value = bb.width;
+    }
+    onMounted(() => onResize());
+    watchEffect(() => {
+      const svg = select('svg');
```

Comment:
This is interesting. I ran into problems when I defined an `svgRef` where calling `select(svgRef.value)` in `watchEffect` before the component was mounted would cause errors because the ref wasn't bound yet. I solved it by doing `onMounted(() => watchEffect(...));`. You use `select('svg') here, which I assume somehow works prior to component mounting? I'm surprised this works as is.

Reply:
I changed it to use an `svgRef` variable inside `onMounted()` like you have in the Scree Plot to make things more uniform.